trio/tests/test_util.py

```diff
@@ -58,3 +61,54 @@ def handler(signum, _):
                 async with ul1:
                     pass  # pragma: no cover
     assert "ul1" in str(excinfo.value)
+
+
+async def test_contextmanager_do_not_unchain_non_stopiteration_exceptions():
+    @acontextmanager
+    @async_generator
+    async def test_issue29692():
+        try:
+            await yield_()
+        except Exception as exc:
+            raise RuntimeError('issue29692:Chained') from exc
+
+    with pytest.raises(RuntimeError) as excinfo:
+        async with test_issue29692():
+            raise ZeroDivisionError
+    assert excinfo.value.args[0] == 'issue29692:Chained'
+    assert isinstance(excinfo.value.__cause__, ZeroDivisionError)
+
+    # XXX: Broken for StopIteration???
+    for cls in [StopIteration, StopAsyncIteration]:
+        with pytest.raises(cls) as excinfo:
+            async with test_issue29692():
+                raise cls('issue29692:Unchained')
+        assert excinfo.value.args[0] == 'issue29692:Unchained'
+        assert excinfo.value.__cause__ is None
+
+
+# Native async generators are only available from Python 3.6 and onwards
+nativeasyncgenerators = sys.version_info >= (3,6)
+pytestmark = pytest.mark.skipif(not nativeasyncgenerators, reason="Python with native async generators")
+
+
+async def test_native_contextmanager_do_not_unchain_non_stopiteration_exceptions():
+    @acontextmanager
+    async def test_issue29692_2():
+        try:
+            yield
+        except Exception as exc:
+            raise RuntimeError('issue29692:Chained') from exc
```

Comment:
Unfortunately 3.5 blows up if you even have syntax like this in your *file*, it doesn't matter whether you try executing it or not. So the only way I know to write a test like this is to use `exec`:
```python
code = textwrap.dedent("""
    @acontextmanager
    async def test_issue29692_2():
        ...
""")
ns = {"acontextmanager": acontextmanager}
exec(code, ns)
test_issue29692_2 = ns["test_issue29692_2"]
```
(and then you still need a version check to make sure you only do this on 3.6+, or alternatively try it always and then catch the `SyntaxError` if it doesn't work)

It might make sense to consolidate with the previous test... like make a version dependent list of test context managers, and then do a `for` loop over it? Or not, sometimes it's simpler just to have a bit of copy-paste in test code :-)